packages/v2/contracts/ANTv2.sol

```diff
@@ -0,0 +1,159 @@
+// Adds:
+//   - A minting role
+//   - transferWithAuth()
+contract ANTv2 is IERC20 {
+    using SafeMath for uint256;
+
+    string private constant ERROR_NOT_MINTER = "ANTV2:NOT_MINTER";
+    string private constant ERROR_AUTH_EXPIRED = "ANTV2:AUTH_EXPIRED";
+    string private constant ERROR_AUTH_NOT_YET_VALID = "ANTV2:AUTH_NOT_YET_VALID";
+    string private constant ERROR_AUTH_ALREADY_USED = "ANTV2:AUTH_ALREADY_USED";
+    string private constant ERROR_INVALID_SIGNATURE = "ANTV2:INVALID_SIGNATURE";
+
+    string public constant name = "Aragon Network Token";
+    string public constant symbol = "ANT";
+    uint8 public constant decimals = 18;
+
+    address public minter;
+    uint256 public totalSupply;
+    mapping (address => uint256) public balanceOf;
+    mapping (address => mapping(address => uint256)) public allowance;
+
+    bytes32 public DOMAIN_SEPARATOR;
+
+    // bytes32 public constant PERMIT_TYPEHASH = keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 public constant PERMIT_TYPEHASH = 0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+    mapping (address => uint256) public nonces;
+
+    // bytes32 public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH =
+    //     keccak256("TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)");
+    bytes32 public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH = 0x7c7c6cdb67a18743f49ec6fa9b35f50d52ed05cbed4cc592e13b44501c1a2267;
+    mapping (address => mapping (bytes32 => bool)) public authorizationState;
+
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+    event Transfer(address indexed from, address indexed to, uint256 value);
+    event AuthorizationUsed(address indexed authorizer, bytes32 indexed nonce);
+    event ChangeMinter(address indexed minter);
+
+    modifier onlyMinter {
+        require(msg.sender == minter, ERROR_NOT_MINTER);
+        _;
+    }
+
+    constructor(uint256 chainId, address initialMinter) public {
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'),
+                keccak256(bytes(name)),
+                keccak256(bytes('1')),
+                chainId,
+                address(this)
+            )
+        );
+
+        _changeMinter(initialMinter);
+    }
+
+    function _validateSignedData(address signer, bytes32 encodeData, uint8 v, bytes32 r, bytes32 s) internal view {
+        bytes32 digest = keccak256(
+            abi.encodePacked(
+                "\x19\x01",
+                DOMAIN_SEPARATOR,
+                encodeData
+            )
+        );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        require(recoveredAddress != address(0) && recoveredAddress == signer, ERROR_INVALID_SIGNATURE);
```

Comment:
Maybe we can leave a comment here on why we need to check with `address(0)` as it's explained in the spec.